src/pool/pool_proxy.c

```diff
@@ -0,0 +1,125 @@
+/*
+ *
+ * Copyright (C) 2024 Intel Corporation
+ *
+ * Under the Apache License v2.0 with LLVM Exceptions. See LICENSE.TXT.
+ * SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+ *
+ */
+
+#include <umf/memory_pool_ops.h>
+#include <umf/pools/pool_proxy.h>
+
+#include <assert.h>
+#include <stdlib.h>
+
+#include "utils_common.h"
+
+static __TLS umf_result_t TLS_last_allocation_error;
+
+struct proxy_memory_pool {
+    umf_memory_provider_handle_t hProvider;
+};
+
+static umf_result_t
+proxy_pool_initialize(umf_memory_provider_handle_t hProvider, void *params,
+                      void **ppPool) {
+    (void)params; // unused
+
+    struct proxy_memory_pool *pool = malloc(sizeof(struct proxy_memory_pool));
+    if (!pool) {
+        return UMF_RESULT_ERROR_OUT_OF_HOST_MEMORY;
+    }
+
+    pool->hProvider = hProvider;
+    *ppPool = (void *)pool;
+
+    return UMF_RESULT_SUCCESS;
+}
+
+static void proxy_pool_finalize(void *pool) { free(pool); }
+
+static void *proxy_aligned_malloc(void *pool, size_t size, size_t alignment) {
+    assert(pool);
+
+    void *ptr;
+    struct proxy_memory_pool *hPool = (struct proxy_memory_pool *)pool;
+
+    umf_result_t ret =
+        umfMemoryProviderAlloc(hPool->hProvider, size, alignment, &ptr);
+    if (ret != UMF_RESULT_SUCCESS) {
+        TLS_last_allocation_error = ret;
+        return NULL;
+    }
+
+    TLS_last_allocation_error = UMF_RESULT_SUCCESS;
+    return ptr;
+}
+
+static void *proxy_malloc(void *pool, size_t size) {
+    assert(pool);
+
+    return proxy_aligned_malloc(pool, size, 0);
+}
+
+static void *proxy_calloc(void *pool, size_t num, size_t size) {
+    assert(pool);
+
+    (void)pool;
+    (void)num;
+    (void)size;
+
+    // Currently we cannot implement calloc in a way that would
+    // work for memory that is inaccessible on the host
```

Comment:
this is specific for some providers - why can't we rely on providers for this? What if I want to use a proxy provider for an OS provider?

Reply:
How do you propose to do that? Providers do not expose calloc or realloc functions.

Reply:
Yes, right. But I think this could be confusing for someone used to malloc API but not having the knowledge about the providers. I think we should add this to the documentation.